inference-engine/src/mkldnn_plugin/nodes/mkldnn_interpolate_node.cpp

```diff
@@ -0,0 +1,1950 @@
+            uint8_t *out_ptr = out_ptr_ + (OW * OH * OD * CB * blk_size * b) * dstDataSize;
+            std::vector<int> index_w_kernel(OW);
+            for (int ox = 0; ox < OW; ox++) {
+                index_w_kernel[ox] = index_w[ox] * blk_size * srcDataSize;
+            }
+            parallel_for2d(CB, OD, [&](size_t cb, size_t d) {
+                uint8_t *out_ptr_cbd = out_ptr + (blk_size * OW * OH * OD * cb + blk_size * OW * OH * d) * dstDataSize;
+                const uint8_t *in_ptr_cbd = in_ptr + (blk_size * IW * IH * ID * cb + blk_size * IW * IH * index_d[d]) * srcDataSize;
+                auto arg = jit_interpolate_call_args();
+                for (int h = 0; h < OH; h++) {  // kernel for blk_size * OW
+                    arg.dst = out_ptr_cbd + blk_size * OW * h * dstDataSize;
+                    arg.src = in_ptr_cbd + blk_size * IW * index_h[h] * srcDataSize;
+                    arg.index = static_cast<int*>(&(index_w_kernel[0]));
+                    arg.work_amount = static_cast<size_t>(OW);
+                    arg.oc_off = cb * blk_size;
+                    (*interpolateKernel)(&arg);
+                }
+            });
+        }
+    }  // batch end
+}
+
+void MKLDNNInterpolateNode::NNRef(const uint8_t *in_ptr_, uint8_t *out_ptr_, int B, int C, int ID, int IH, int IW,
+                                          float fx, float fy, float fz, int OD, int OH, int OW) {
+    std::vector<int> index_buffer(OD * OH * OW);
+    bool isDDownsample = (fz > 1) ? true : false;
+    bool isHDownsample = (fy > 1) ? true : false;
+    bool isWDownsample = (fx > 1) ? true : false;
+    int nearestVoxel = 0;
+    for (int oz = 0; oz < OD; oz++) {
+        float iz = coordTransToInput(oz, fz, ID, OD);
+        nearestVoxel = nearestRound(iz, isDDownsample);
+        int iz_offset = nearestVoxel * IH * IW;
+        int oz_offset = oz * OH * OW;
+        for (int oy = 0; oy < OH; oy++) {
+            float iy = coordTransToInput(oy, fy, IH, OH);
+            nearestVoxel = nearestRound(iy, isHDownsample);
+            int iy_offset = nearestVoxel * IW + iz_offset;
+            int oy_offset = oy * OW + oz_offset;
+            for (int ox = 0; ox < OW; ox++) {
+                float ix = coordTransToInput(ox, fx, IW, OW);
+                nearestVoxel = nearestRound(ix, isWDownsample);
+                int ix_index = nearestVoxel + iy_offset;
+                index_buffer[oy_offset + ox] = ix_index;
+            }
+        }
+    }
+
+    parallel_for2d(B, C, [&](size_t b, size_t c) {
+        const uint8_t *in_ptr = in_ptr_ + (IW * IH * ID * C * b + IW * IH * ID * c) * srcDataSize;
+        uint8_t *out_ptr = out_ptr_ + (OW * OH * OD * C * b + OW * OH * OD * c) * dstDataSize;
+
+        for (int i_dst = 0; i_dst < OW * OH * OD; i_dst++) {
+            float dstValue = getValue(in_ptr, index_buffer[i_dst] * srcDataSize, inputPrec);
+            if (!fusedWith.empty()) {
+                applyPostOpsScalar(dstValue, c);
```

Comment:
My thought is: Though here post ops is done without fusion in JIT kernel as attribute of interpolate layer, we post process the data just after we get it can reduce memory pressure in some degrees as well. At the post ops time, the data is stored very close to processor(at least in L1 cache) and read it to processor very quickly. Otherwise we need go through the tensors again for post ops. A more worse case is post ops is assigned to other cores, which let read data more slow. Am I right?

Reply:
I general you are right. However here we have two details that ruins everything:
1) `applyPostOpsScalar` routine is very heavy itself. It contains cycles, conditions, function calls. So we have much more intensive compute.
2) `applyPostOpsScalar` routine cannot be vectorized by compiler. So you will compute by scalar code not only Interpolate but also fused operation. If you disable fusing, these operations can be computed via vector kernels which in general faster even with higher memory pressure.
My experience says that from performance perspective it is better no to use scalar post ops.

Reply:
what you say make sense!
Because "can JITTED" can not determined from `mayiuse(sse42) == ture` directly, due planar layout is jitted with gather instrument, which only available until avx2 and f32. so I am think if I can use `interpolateKernel == nullptr` to judage if jitted in graph opt. interpolateKernel is initialized in createPrimitive(). so my question is: what's the call sequence between `MKLDNNInterpolateNode::createPrimitive() `and `MKLDNNGraphOptimizer::FuseInterpolateAndSimpleOperation` ?

Reply:
MKLDNNGraphOptimizer::FuseInterpolateAndSimpleOperation is called before MKLDNNInterpolateNode::createPrimitive(). You can see the order here: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp#L343

So as I said my suggestion is to create canFuse method. See my PR for reference: https://github.com/openvinotoolkit/openvino/blob/0806523db255c8d823981e0f3a126cb51b5cd480/inference-engine/src/mkldnn_plugin/nodes/mkldnn_eltwise_node.cpp#L1370 and https://github.com/openvinotoolkit/openvino/blob/0806523db255c8d823981e0f3a126cb51b5cd480/inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp#L1871

The simplest way is to copy logic that you have in createPrimitive into canFuse in order to understand will jitted primitive be created or not.
Regarding specialty for layouts - it is not that simple. In general we are trying to implement symmetric jitter support for all cases :). In your particular case I would advise not to create planar config at all in case we have smt fused.

Reply:
Done. Remove all fusion when they can not fused into JIT kernel.
Added canFuse() and it is true when we have sse42. Planar layout config is created when 1: no sse42 for backup implementation. 2: when "can JITTED" condition is satisfied(have avx2 and fp32 input precision). Then if we have fused ops, we can select blk and by_channel. or planar if it's available when `mayiuse(sse42) == true.`